File: code/day2.R
# Day 2 advent of code R solution

psw <- read.table(file.path("data", "day2"))
names(psw) <- c("rule", "target", "pw")

# puzzle 1 ----------------------------------------------------------------

check_psw <- function(rule, target, pw) {
  bnds <- as.numeric(unlist(strsplit(rule, "-")))
  n <- stringr::str_count(pw, sub(":", "", target))
  if (n <= bnds[2] && n >= bnds[1]) return(TRUE)
  return(FALSE)
}

cat("Part 1 answer is:", sum(purrr::pmap_lgl(psw, check_psw)), "\n")

# puzzle 2 ----------------------------------------------------------------

check_psw2 <- function(rule, target, pw) {
  bnds <- as.numeric(unlist(strsplit(rule, "-")))
  loc <- unlist(gregexpr(sub(":", "", target), pw))
  if (sum(bnds %in% loc) == 1) return(TRUE)
  return(FALSE)
}

cat("Part 2 answer is:", sum(purrr::pmap_lgl(psw, check_psw2)), "\n")

File: code/day2.jl
# Julia solution to Day 2 of Advent of Code

using DelimitedFiles

pw = readdlm("data/day2")

# puzzle 1
function check_pw1(rule, target, password)
    bounds = [parse(Int, split(rule, "-")[i]) for i in 1:2]
    count(i -> (i == target[1]), password) in bounds[1]:bounds[2]
end

println(count(i -> (check_pw1(pw[i, 1], pw[i, 2], pw[i, 3])), 1:1000))

# puzzle 2
function check_pw2(rule, target, password)
    bounds = [parse(Int, split(rule, "-")[i]) for i in 1:2]
    count(i -> (password[i] == target[1]), bounds) == 1
end

println(count(i -> (check_pw2(pw[i, 1], pw[i, 2], pw[i, 3])), 1:1000))
